cmd/acra-server/common/listener.go

```diff
@@ -291,8 +318,19 @@ func (server *SServer) ListenerAPI() net.Listener {
 	return server.listenerAPI
 }
 
+func (server *SServer) waitForExitTimeout() {
+	// We should use this function when shutdown service as a defer. In this case global 'cancel'
+	// has been called. Now we should wait (not more than specified duration) until all
+	// background goroutines spawned by SServer will finish their execution or force their closing.
+	if utils.WaitWithTimeout(&server.backgroundWorkersSync, utils.DefaultWaitGroupTimeoutDuration) {
+		log.Errorf("Couldn't stop all background goroutines spawned by readerServer. Exited by timeout")
+	}
+}
+
 // Start listening connections from proxy
-func (server *SServer) Start() {
+func (server *SServer) Start(parentContext context.Context) {
+	defer server.waitForExitTimeout()
```

Comment:
Is this in the right place? Do we need to wait even if, say, `network.Listen(...)` fails? Shouldn't this be just a regular call after `server.run(...)`? Or can it be just a part of `run()`?

Ditto in `StartFromFileDescriptor`, `StartCommands`, `StartCommandsFromFileDescriptor` with their respective failures.

Reply:
nice catch. better to call it after an initialized listener.

Reply:
Got it. Thanks